RFC 149 - Switch to per-page asset loading (#152)
alphagov/govuk-rfcs: rfc-149-switch-to-per-page-asset-loading.md

### Auditing component use in applications is a manual process.

Each application only bundles the CSS and JavaScript for the components used - [see RFC #108 for more information][specific_assets].

Adding a component's CSS and JavaScript to an application's stylesheet and JavaScript is a manual process - which means that it's easy to forget to remove assets when a component has stopped being used in an application. We have auditing tools to help manage this - but they are not part of our deployment pipeline and rely on developers remembering to use them.

Supplying the assets that a component needs as part of the partial would mean that assets would only be included when the component is used - this would simplify the developer experience when adding or removing components from applications, and would also mean a better user experience since unused assets wouldn't be loaded.

## Proposal

> Assets should be included individually on a per-component basis and included in the component partial.

Author: There isn't an overlap between components used in static and the rendering apps - static only does the header and the footer, so only renders (off the top of my head) the skip-link, navigation-header, footer, banners, and feedback components. These can use individual assets and won't clash with the rendering applications as the rendering applications don't use these components.

We manually stick add components that are shared across a lot of the applications to stop the same CSS and JavaScript from being downloaded unnecessarily - this won't be needed in the future, so would stop any mismatched versions causing any breakages.

> Assets should be included individually on a per-component basis and included in the component partial.

Reviewer: So components I'm referencing are things like input, hint, label. Those that are used as part of the components static uses and are then also used in applications.

It's unclear to me how a rendering app would know that any static components used those components (well outside of hacking away at Slimmer)

> Assets should be included individually on a per-component basis and included in the component partial.

Author: Hmm, not sure exactly how this would be done. We could:

* have a 'components used in Static' list in the gem that we check against - which would need maintaining, similar to the current set up
* do any deduping of stylesheets in Slimmer after the two templates have been knitted together - no list needed, but uses Slimmer which isn't ideal

I'd prefer the first option as it has no reliance on Slimmer - but any other suggestions welcome.

The other thing worth noting is that this stops becoming a problem when Static/Slimmer is not longer being used - though that's probably a while off!

> Assets should be included individually on a per-component basis and included in the component partial.

Reviewer: Yup, I think the first option is preferable. Using slimmer makes me uncomfortable as it's very difficult to test and will likely be fragile.

> The other thing worth noting is that this stops becoming a problem when Static/Slimmer is not longer being used - though that's probably a while off!

I dream of the day!

> This would mean that a page loads `component.css` and `component.js` only when a component is present on the page - rather than the current behaviour, which sees component assets loaded if used anywhere within an application. If multiple uses of the same component occur on the page, there should only be one instance of the assets included on the page.

Author: Dependencies is a tricky one; JavaScript is a bit of an unknown as there are other factors outside of the RFC at play.

a) For JavaScript, I'm not completely sure. What I hope is that with the potential to drop JavaScript support for Internet Explorer 11 we can investigate the use of ES6 modules. This means that including polyfills and dependencies becomes a lot easier: `import { dependency } from 'dependency'`. Because the import is path based, different component's JavaScript can use the same dependency without it being downloaded multiple times - and they're only downloaded when needed rather than being bundled up whether they're needed or not. (We can look at doing even nicer things like only importing modules when a condition is met - so polyfills can be downloaded only when needed.)

For CSS, we'd include everything needed from GOVUK Frontend - apart from components - into one file and serve that; this should include all the base styles (`govuk-body`, `govuk-heading-*` etc) that are used on every single page. (This is more or less what `govuk_frontend_support` and `component_support` is.) Having this in its own file will mean that it will be more likely to be cached by the browser.

b) There's an agreement that we don't use the Sass `@extend .classname` feature as it creates too much junk CSS - it's smaller and simpler to just use two classes in the HTML. This means that we don't need access to the full GOVUK Frontend Sass in each component's stylesheet.

The common mixins would be included if needed by `@import`ing the required bits from GOVUK Frontend - either by using `gouvk_frontend_support` or a subsection of GOVUK Frontend itself. Some components would need this, some wouldn't. This won't add anything extra to the component's stylesheet.

The lack of `@use` is a problem - this is exacerbated by the lack of `@use` support in the version of Sass that we use. Moving to Dart Sass would help build the case for GOVUK Frontend dropping libsass and RubySass support - so there is a future solution to this, but not one yet.

c) Yes, an `all_components` Sass file should still be supported if needed - it'd be good to know where this is used though and how much of a need there is for it.

`govuk_frontend_support` is only mixins and variables - it doesn't output any CSS, so can be `@import`ed into every component without adding any extra size. I'm measuring how much impact this has when developing locally at the moment with the gem and the `frontend` rendering app - I do worry about the developer experience, but until we can acertain the level of impact I don't think that this should block this proposal.

`component_support` does output CSS - so it's only used in Static on the public facing aspect of GOVUK. This should be part of the previously mentioned stylesheet that has all the basic GOVUK Frontend stuff in it.

> This would mean that a page loads `component.css` and `component.js` only when a component is present on the page - rather than the current behaviour, which sees component assets loaded if used anywhere within an application. If multiple uses of the same component occur on the page, there should only be one instance of the assets included on the page.

Reviewer: Sure, that sounds reasonable sounds like there's a fair amount of unknown.

With JS, would we launch the individual files when we've got a similar set-up for JS and CSS. I.e individual files for each. Or would this propose doing just CSS and then leaving JS to later (so application.js entries for JS but individual files for CSS). I'm kinda hoping it'd be the former.

One of the key places of all_components is in the component-guide itself, otherwise I think it's in backend apps that haven't had much love recently (Content Publisher, Content Data etc) 

If this work brings us closer towards Dart and/or modernising the JS that'd be a good outcome

> This would mean that a page loads `component.css` and `component.js` only when a component is present on the page - rather than the current behaviour, which sees component assets loaded if used anywhere within an application. If multiple uses of the same component occur on the page, there should only be one instance of the assets included on the page.

Author: We'd do this for CSS first, and then tackle the JavaScript afterwards - so, as you say, there'd be `application.js` and individual CSS files. 

Whilst it's not perfect, even just serving the CSS individually would be an improvement - and I worry that doing both at the same time would be too big a chunk of work to get done in a sensible time frame.

Waiting for clearer direction for IE11 support, the upcoming translation work from GOVUK Frontend, and investigation into how the latest version of Rails uses ES6 modules would mean a longer wait until the JavaScript is loaded individually, but potentially less work to get it done.

> This would mean that a page loads `component.css` and `component.js` only when a component is present on the page - rather than the current behaviour, which sees component assets loaded if used anywhere within an application. If multiple uses of the same component occur on the page, there should only be one instance of the assets included on the page.

Reviewer: Ahh sure, that does seem somewhat messy as we'd have a separate install step for the JS and stylesheet. 

It seems ok to go with that as a first step and agree with your point about sensible time frame, but I really wouldn't wait for ES6 modules as we've got so many things to resolve to get to that. I imagine we could just create an equivalent JS file of component support and then just add each component as extra JS, similar to CSS. 

As an aside, I do wonder how we can start work on what JS without IE support looks like and the problems to overcome.

> ### Keep the stylesheets in the head

Author: Moved in https://github.com/alphagov/govuk-rfcs/pull/152/commits/779531b0f023733da76bc92c5758697560d95514.

A version identifier should be added to the component asset filename to help identify which version of the components gem is being used - for example `gem-v29-2-1-accordion-09288...adb36.css`. The fingerprint should remain to show that the contents of the file are the same.

The bespoke app assets should be renamed to their application name - so collections would have `collections-5f801d...179b0.css` and `collections-828a8...abb6c.js`. This will make it easier to see which stylesheet comes from which application.

Because each application is using the components gem as the source for the assets for each component, the compiled assets will be the same - whether the resulting CSS or JavaScript file is produced by whitehall, frontend, or collections. When writing to the shared folder the component assets may be overwritten, but - thanks to the fingerprint - they'll be overwritten with files containing the same content.

> ## Considerations

Author: I think that this should be an optional choice and backwards compatible - so if that means running a `accordion-individual.scss` file that `@import`s from `govuk_frontend_support` and from the `_accordion.scss` then that's okay. I'm aware that there are applications that don't want this approach, and the gem needs to work (and work without taking a long time) for both publishing and rendering applications.

A quick check of frontend running using app-live in GOVUK Docker, cold start:

* using gem: 17 seconds
* using local version of the gem: 44 seconds
* using local version of the gem with `@import govuk_frontend_support` for each component that requires it: 59 seconds

Whilst not great, I think it's an acceptable increase whilst there are other avenues worth investigating to make this faster - off the top of my head, more efficient ways of only importing the specific things that are needed from GOVUK Frontend, using the faster Dart Sass, use of `@use` within the components gem and rendering applications.

This is only a single run, so I'll do some more investigation into this tomorrow to be able to report an average number of runs. 

Whilst not a complete solution, running it using `./startup.sh --live` seems to be about four times faster, which makes me suspicious that I'm not doing a completely cold start.

I'd suggest that an unacceptable regression would be if a cold start is above a minute under the current libsass and Docker set up.

> ## Considerations

Reviewer: Great. Yup having files without needing the import sounds great to avoid performance penalties everywhere. If it's a 25% increase as per your experiment that isn't too bad either - it's an order of magnitude better than I feared.

A minute sounds good too, so long as we have a certain point where we're prepared to go "eep maybe we should focus on getting dart in first" works for me

Yeah to do a full cold start you need to clear the cache, `rails assets:clobber tmp:clear` I think to be complete

Of the top 100 most visited pages on GOV.UK, the average number of components on a page is 19. The smallest number of components on a page is 9, and the largest number of components is 25. It's worth noting that most - but not all - components need a stylesheet (72 out of 78 components have a stylesheet) and not every component needs JavaScript (24 out of 78 components have a JavaScript file) - so adding a component to a page doesn't always mean two more requests are made.

It's worth noting that this change will lead to slower load times for browsers that only support HTTP/1.1 - even when considering the reduction in page size. ([More information on HTTP/1.1's limitations][http_limitations]). This will impact Internet Explorer 11 on Windows 8 and below; the first versions of Chrome, Firefox, and Opera that supported HTTP/2 were released in early 2015.

This balance between making the site load faster for a majority of users is not just a numbers game - the users who are most likely to be using older browsers are those that are likely to least afford the upgrade cost for a newer device, and they're likely to be the users who most need the information on GOV.UK.

> Whilst this proposed change will make things slower for users whose browser doesn't support HTTP/2 or HTTP/3, it won't prevent pages from loading. This proposed change should be accepted because it will dramatically improve page performance for almost all visitors without blocking access for a small number of visitors.

Author: I've included how much less CSS will be downloaded on the top 100 pages - all of the pages showed a reduction in the amount of CSS serve.

But since the other part of this is to improve the cacheability of the files I don't think we'll be able to tell that until we've look at the real user metrics. A bit of a cop out I'm afraid!

If you're on a fast-ish wired connection you probably won't notice any difference. If you're on a mobile connection it should really help, especially if it's flaky with a risk of both dropped packets and connections.

## Summary

Component stylesheets and JavaScript should only be included for components present on a page on GOV.UK. This will lead to a reduction in page size, improve the ability of the browser to cache assets - which in turn improves performance for user journeys and return visits - and help reduce our reliance on manual asset auditing.

The eleven applications that render GOV.UK use the [GOV.UK Publishing Components][govuk_publishing_components] gem. This gem is the single source of truth for shared components used across GOV.UK.

Users that visit more than one page are likely to see pages rendered by different applications. Applications have unique but overlapping combinations of component use - this means that a user could download the same code in each application's concatenated stylesheets and JavaScript files when visiting pages rendered by different applications, despite the components being the same.

Because each application concatenates and appends a fingerprint string to an asset's filename, even a small change to a component will cause the filename to change - meaning that return visitors will no longer have the advantage of caching.

> Currently asset auditing is a manual process that needs to be run locally. This means it's easy to miss when a component is no longer used, which leads to larger than needed stylesheets and JavaScript files being served to users.

Author: So the summary would just be the following and nothing else?

> Component stylesheets and JavaScript should only be included for components present on a page on GOV.UK. This will lead to a reduction in page size, improve the ability of the browser to cache assets - which in turn improves performance for user journeys and return visits - and help reduce our reliance on manual asset auditing.

> Currently asset auditing is a manual process that needs to be run locally. This means it's easy to miss when a component is no longer used, which leads to larger than needed stylesheets and JavaScript files being served to users.

Author: Done in https://github.com/alphagov/govuk-rfcs/pull/152/commits/b97a09a5957c7c70f46ca66e8eee4dc318a1c059.

This RFC builds upon ideas and implementations from:
* [RFC #91: Sharing assets][sharing_assets] (not implemented)
* [RFC #108: Include specific component assets in applications][specific_assets] (implemented)
* [RFC #115 : Enabling HTTP/2 on GOV.UK][http2] (implemented)

## Problems

### User journeys that visit more than one rendering app are inefficient

Four out of ten tracked visitors to GOV.UK in the past year have visited more than one page in a single session.

Each of the eleven rendering applications have different sets of CSS and JavaScript files containing both code specific to that application and code coming from shared components.

These four out of ten visitors are likely to have visited pages rendered by multiple applications - which means they have downloaded the same code more than once via different files.

For example, a three step journey can visit three different rendering applications:

1. Homepage, rendered by [frontend]
1. Search for "micro pig", rendered by [finder-frontend]
1. Micro-pig guidance page, rendered by [government-frontend]

> Whilst there is shared code coming from static, CSS and JavaScript can still come from each individual application - which means there is overlap between the three application's assets. In the previous three step journey example, there are 43 component's assets served; of which:

Reviewer 2: Nitpick:

```suggestion
Whilst there is shared code coming from [static], CSS and JavaScript can still come from each individual application - which means there is overlap between the three applications' assets. In the previous three step journey example, there are 43 component's assets served; of which:

```

The component gem is released regularly - over the past year it's been released on average once every four days, but most releases were between one and seven days apart. Two thirds of repeat visitors to GOV.UK return within seven days and one quarter of repeat visitors return between 8 to 30 days. This means it is highly likely that returning visitors need to download the asset files again, regardless of how large or small any changes are.

### Auditing component use in applications is a manual process.

Each application only bundles the CSS and JavaScript for the components used - [see RFC #108 for more information][specific_assets].

Adding a component's CSS and JavaScript to an application's stylesheet and JavaScript is a manual process - which means that it's easy to forget to remove assets when a component has stopped being used in an application. We have auditing tools to help manage this - but they are not part of our deployment pipeline and rely on developers remembering to use them.

> Supplying the assets that a component needs as part of the partial would mean that assets would only be included when the component is used - this would simplify the developer experience when adding or removing components from applications, and would also mean a better user experience since unused assets wouldn't be loaded.

Reviewer 2: This is a big win 👏

[Slimmer][slimmer] currently takes all `script` elements that don't have `defer` or `async` attributes and [moves them to the end of the markup][slimmer_moves_scripts]. This ensures that they are parsed and executed in a known order when the DOM is ready to be read. This behaviour shouldn't be changed.

### Shared assets

The assets should be placed in a folder that allows each application to reference them. The current behaviour has each application place their assets in a different folder - `/assets/collections`, `/assets/finder-frontend` etc. This prevents the browser cache from being used effectively as `/assets/collection/accordion.css` will be downloaded even if `/assets/finder-frontend/accordion.css` has already been downloaded.

A version identifier should be added to the component asset filename to help identify which version of the components gem is being used - for example `gem-v29-2-1-accordion.css`. Similarly, the bespoke app assets should be renamed to their application - so collections would have `collections.css` and `collections.js` - to make it easier to see which stylesheet comes from which application.

> The assets would ideally be placed in a shared folder as per [RFC #91][sharing_assets], so would be able to be used across GOV.UK irrespective of the rendering application. Since this hasn't been implemented yet, there would need to be some work for this to happen - and likely this would be delayed to coincide with the replatforming work.

Author: I'll rewrite this to be clearer - but the tl;dr is that post RFC-91 and RFC-149 the asset URLs should look like:

* `www.gov.uk/assets/gem-v29-2-1-accordion-e03sv...98hg5.css`
* `www.gov.uk/assets/government-frontend-00d69...a20c7.css`

(The assets are currently served off of the `www.gov.uk` domain since [HTTP/2 was enabled](https://github.com/alphagov/govuk-rfcs/blob/main/rfc-115-enabling-http2-on-govuk.md#proposal).)

The renaming of the `application-*.css` is because I think (though am happy to be wrong) that Rails can only use a single asset folder, which is set using `config.assets.prefix`[^1]:

```ruby
    # Path within public/ where assets are compiled to
    config.assets.prefix = "/assets/frontend"
``` 

If this is correct, then a Rails app wouldn't be able to use assets from `/shared` and `/government-frontend` at the same time.

Since all the applications use `application-*.css` and `application-*.js` for their bespoke assets, there'd be a lot of similarly named files where the only difference is the fingerprint - which isn't that useful when trying to tell which file is used by which application. Setting the file name to include the application name seems a good solution to that.

I'd prefer to keep the fingerprint as part of the individual assets to make doubly sure that each file with the same name has the same contents - for example, the same Sass file compiled with Dart-Sass would give a different fingerprint to one compiled with Node-Sass because they have different default number of decimal places[^2].

[^1]: Example taken from [alphagov/frontend/config/application.rb](https://github.com/alphagov/frontend/blob/3ccc1bf52c980852eca61250727f81034930519d/config/application.rb#L58-L59).
[^2]: There's an [interesting issue on how this affects GOV.UK Frontend](https://github.com/alphagov/govuk-frontend/issues/2199).

> The assets would ideally be placed in a shared folder as per [RFC #91][sharing_assets], so would be able to be used across GOV.UK irrespective of the rendering application. Since this hasn't been implemented yet, there would need to be some work for this to happen - and likely this would be delayed to coincide with the replatforming work.

Reviewer 2: Thanks - I think I mostly follow. Can we look at an example just to iron out the last of the confusion?

If we take Frontend as an example and look at its [use of the tabs component](https://github.com/alphagov/frontend/blob/09b27f918b987afb6b1f6dddfdbfb8961dbadcd0/app/views/calendar/bank_holidays.html.erb#L69), it currently [imports the SCSS](https://github.com/alphagov/frontend/blob/6ec5131c8edeceff4e60b118576482be805a9e40/app/assets/stylesheets/application.scss#L35) and [JS](https://github.com/alphagov/frontend/blob/8dbe2e4de266be77da511156f476fb52b2b09166/app/assets/javascripts/application.js#L7) manually, combining it with all of the application assets and outputting as single files `/assets/frontend/application-FINGERPRINT.css` and `/assets/frontend/application-FINGERPRINT.js` respectively.

For stage 1, we want to implement the individual serving of assets. This would involve:

1. Removing the `import`/`require` calls to the component
2. Pulling in the component CSS/JS in some other way - **this is where I'm stuck**. It may be an implementation detail but it could end up changing parts of the proposal, so worth digging a little deeper I think. I'm not clear from the RFC how this would be done.

I think what you're proposing is roughly this:

> Write a compile-time 'hook' into the `render` method, which copies the component CSS/JS files into the Rails project, as separate files, alongside the application asset source files.
> This will then output as `/assets/frontend/gem-v29-2-1-tabs-FINGERPRINT.css` etc. We rely on Slimmer to yank the `<link rel` element and move it into the head.
> The component template itself makes no reference to assets.
>
> Then for stage 2, it's just a case of changing the `config.assets.prefix` value to something like `/assets/shared` and renaming `application.scss` and `application.js` to `frontend.scss|js`. _Now_ the component would be uploaded to `/assets/shared/gem-v29-2-1-tabs-FINGERPRINT.css` instead.
> And if another app (e.g. Whitehall) uses the same version of the same component, it will also output to the same place, essentially overwriting the contents, but we don't care since the contents (and therefore the fingerprint) would be the same. So at that point, Frontend is effectively using the tabs assets compiled by Whitehall, but there is no difference in the bytes being sent to the user.

Does that make sense? Have I understood that right?

The other alternative I can think of is to have the component templates refer to their own assets, like in the pseudocode below, but I can't think of a way of accomplishing that whilst retaining the fingerprint, and it would also need a separate build pipeline (likely a GitHub Action).

```erb
<% if use_individual_assets %>
  <link rel="stylesheet" href="<%=`/assets/shared/gem-v${VERSION}-tabs.css` %>" />
<% end %>
``` 

(Might make the 'optional' logic easier though - just a case of passing `use_individual_assets: false` at the application level).

> The assets would ideally be placed in a shared folder as per [RFC #91][sharing_assets], so would be able to be used across GOV.UK irrespective of the rendering application. Since this hasn't been implemented yet, there would need to be some work for this to happen - and likely this would be delayed to coincide with the replatforming work.

Author: Yes, that's what I think should happen. There's some discussion that should happen during the implementation about exactly how and where things should happen - but that's getting a bit too much into the weeds. 

From digging into how the component gem interacts with the rendering applications I can't find anything that screams that this isn't doable. GitHub have started to break their CSS bundle up into more managable chunks - [there was a presentation at this years RailsConf on it](https://youtu.be/21QG19Zy_g0?t=989) ([slides](https://hawksley.org/2022/03/01/conference-proposal-breaking-up-with-the-bundle.html)) - which also makes me think that this is a achievable.

> The assets would ideally be placed in a shared folder as per [RFC #91][sharing_assets], so would be able to be used across GOV.UK irrespective of the rendering application. Since this hasn't been implemented yet, there would need to be some work for this to happen - and likely this would be delayed to coincide with the replatforming work.

Author: I've rewritten the section in https://github.com/alphagov/govuk-rfcs/pull/152/commits/adbc60a9ab8861811d3098f5a4c7dea78b552681 - does it make more sense?

This work should not block the serving of individual assets as there are advantages that don't rely on having a shared folder.

## Considerations

### Reduced compressibility when serving individual files

GOV.UK uses [Brotli compression][enable_brotli] when serving assets - this is a compression method that works really well on repetitive text like HTML, CSS, and JavaScript. Serving individual files will reduce the amount of compression available because the files will have less repetition in them. Since each page will only be serving the component assets required on the page - rather than the components used in the entire application - there should be a reduction in asset size.

To determine whether this is correct, I looked at the difference in CSS size for the top 100 most visited pages on GOV.UK over the past year (1 July 2021 to 31 June 2022). There was reduction in CSS size for pages[^1].

The largest change in component CSS size is -12.2kB, the smallest is -1kB, and the average decrease is 5.8kB; the graph shows the top 100 pages and the reduction in the amount of CSS for each page:

![](rfc-149/css-size-for-top-100-most-visited-pages-on-govuk.svg)

> This means that all visitors will see a smaller page size, regardless of whether they visit one page or multiple pages. Six out of ten tracked visitors to GOVUK only view a single page.
> 
> Concatenating the assets can provide a benefit for those that visit multiple pages - but this doesn't work on GOVUK due to the number of rendering apps used and the current lack of a shared space to put assets.

Reviewer 2: ```suggestion
This means that all visitors will see a smaller page size, regardless of whether they visit one page or multiple pages. Six out of ten tracked visitors to GOV.UK only view a single page.

Concatenating the assets can provide a benefit for those that visit multiple pages - but this doesn't work on GOV.UK due to the number of rendering apps used and the current lack of a shared space to put assets.
```

Author: Again, good spot! Fixed in https://github.com/alphagov/govuk-rfcs/pull/152/commits/35f516e8f71c2a0dc69e67367f0ac35056a6d722

This balance between making the site load faster for a majority of users is not just a numbers game - the users who are most likely to be using older browsers are those that are likely to least afford the upgrade cost for a newer device, and they're likely to be the users who most need the information on GOV.UK.

Whilst this proposed change will make things slower for users whose browser doesn't support HTTP/2 or HTTP/3, it won't prevent pages from loading. This proposed change should be accepted because it will dramatically improve page performance for almost all visitors without blocking access for a small number of visitors.

### Non-rendering applications that use the components gem

There are applications that use the components gem, but don't render public pages on GOV.UK - for example, publishing applications and the Govspeak gem. These may have a different set up with different ways of rendering pages.

> These changes should be optional - even if this is the default behaviour, it should be able to be turned off at a component by component level and at an application level. This could be done with a helper method that checks both if the component has already been used, and if adding the component assets are allowed:

Reviewer 2: This adds a fair bit of complexity. Is there an advantage to making things optional?
Do any applications use govuk_publishing_components without Static / Slimmer?

Author: The publishing applications and the govspeak gem use the components gem without Static - I'll made it clearer in the previous paragraph.

Author: Added in https://github.com/alphagov/govuk-rfcs/pull/152/commits/64480aac48b0feff8489c7036525c546a4b079ef

```ruby
# component_asset_helper.rb - pseudo code example:

def component_assets_allowed (component_name)
  # check:
  #  - if the component has already had the assets added to the page
  # - if the component assets loading method is allowed at an application level
  # - if the individual component has used an option to turn individual assets on or off
  @components_in_use.include?(component_name) || ENV.component_asset_loading !== "bulk" || use_individual_assets === true
end
```

## Other solutions that have been considered and rejected

* Put all components used across GOV.UK in one stylesheet and one JavaScript file - this increases page weight and doesn't solve cache invalidation
* Serve one stylesheet and one JavaScript file for all of GOV.UK - this increases page weight and doesn't solve cache invalidation
* Serve one stylesheet and one JavaScript file per template - this doesn't solve cache invalidation when one component changes

> * Serve all component assets used in an application in individual files - this increases page weight for single page visitors, though often decreased it for people who visited more than one page

Author: This is about serving all individual assets for an application, rather than for the page being viewed - so a page rendered by collections would include all of the assets needed for the entire app, rather than just that page. Kind of like the current behaviour, except with the assets loaded individually instead of being concatenated.

## Implementation process

Serving assets individually doesn't rely on the shared asset folder proposed in RFC 91 - so the lack of a shared folder is not a blocker.

* implement the individual serving of assets
   1. Stylesheets
   1. JavaScript, as this will require more investigation and relies on external factors
* once a shared folder is available, make all applications load assets from the shared folder

### Implement the individual serving of assets

Splitting this into two stages makes it more manageable and will allow improvements to be in stages.

> The stages themselves can be split into app-by-app upgrades - once an app is upgraded the pages it renders will see performance improvements without relying on any other app also being upgraded. This is a similar approach to the work that was done to use the public layout component, so we know it can work - even if it takes time.

Author: Hmm, I'm not sure there is anything written down about the roll out of the public layout component. There are individual pull requests to update each application, so that could show the time between the first and last app to be upgraded - do you think that would be useful?

> The stages themselves can be split into app-by-app upgrades - once an app is upgraded the pages it renders will see performance improvements without relying on any other app also being upgraded. This is a similar approach to the work that was done to use the public layout component, so we know it can work - even if it takes time.

Reviewer 2: A link to one of the PRs sounds good 👍 i.e. `This is a similar approach to the work that was done to use the public layout component ([example](#)), so we know it can work`.

> The stages themselves can be split into app-by-app upgrades - once an app is upgraded the pages it renders will see performance improvements without relying on any other app also being upgraded. This is a similar approach to the work that was done to use the public layout component, so we know it can work - even if it takes time.

Author: I've added examples from near the start of the work (collections) and the end (whitehall of course) in https://github.com/alphagov/govuk-rfcs/pull/152/commits/07022a4f09ea4b3295ccd0c5273a5e31d5ed57e0 - is this okay?

> The stages themselves can be split into app-by-app upgrades - once an app is upgraded the pages it renders will see performance improvements without relying on any other app also being upgraded. This is a similar approach to the work that was done to use the public layout component, so we know it can work - even if it takes time.

Reviewer 2: Brilliant, thanks 👍

This shouldn't be done on a component by component basis - changing only one component on a page will likely cause an increase in page size as the individual component assets can't take advantage of compression.

There are more unknowns with the JavaScript implementation - and this aspect of the work may require further investigation and changes to applications.

It's worth noting that ES6 modules and import maps are the [direction of travel that Ruby on Rails is taking][rails_direction], so this could be the path of least resistance that this implementation could take. GOV.UK Frontend also offers module based imports now, and has plans to drop JavaScript support for Internet Explorer 11 in the near future. It'd be worth seeing how close to a decision these are before choosing a direction.

### Make assets load from a shared folder

> Parallel to this, we need to see how RFC 91 could be implemented. This is complicated by the replatforming project nearing completion.

Reviewer 2: Is it? I'm not sure that replatforming will change the asset compilation/upload pipeline 🤔 I think any changes we make in this area would be carried over automatically - though @sengi will be able to confirm.

Author: Because RFC 91 hasn't been implemented there is no shared folder that shared assets could be uploaded to.

We could implement a shared folder for the current set up - but from discussions it seems likely that this would be finished just in time for replatforming to be switched on.

Possibly complicated is the wrong word - I'm try to note that the shared folder is an external factor with its own RFC, and finding out how this should be carried out is a piece of work in itself.

Reviewer 3: In the new system, Rails assets will be uploaded to a common S3 bucket on rollout (i.e. when a version of an app is deployed to the cluster). It should be feasible to share assets. Right now we're still serving files via the cluster instead of going straight to S3 from the edge (Fastly), but we'd like to do that if not before launch then soon after.

Reviewer 3: https://docs.google.com/drawings/d/1UjW5El9AnrqzgdXjAXKfYBNFR0xNth54hT1PvnMsPxc/edit

Reviewer 3: We should have a more concrete time frame soon but we're talking this year.

Author: That's super news @sengi! Apologies, I was unaware that a shared folder was being implemented so I incorrectly assumed that it wasn't part of the new system.

I'll update this RFC to add that a shared folder is coming soon, and what this will mean for this proposal.

Author: I've updated it to make it clearer that the shared folder is coming soon - https://github.com/alphagov/govuk-rfcs/pull/152/commits/09a5bd268a01f73da3ab6b56da309b03652a5497. Does this sound correct?

[http2]: https://github.com/alphagov/govuk-rfcs/blob/95b4f967a43b24141c4cd0c7feb37f3c309e21c8/rfc-115-enabling-http2-on-govuk.md
[http3]: https://github.com/alphagov/govuk-rfcs/blob/95b4f967a43b24141c4cd0c7feb37f3c309e21c8/rfc-139-enable-http3.md
[link_stylesheet_is_body_okay]: https://html.spec.whatwg.org/multipage/links.html#link-type-stylesheet
[prerelease]: https://guides.rubygems.org/patterns/#prerelease-gems
[rails_direction]: https://world.hey.com/dhh/modern-web-apps-without-javascript-bundling-or-transpiling-a20f2755
[sharing_assets]: https://github.com/alphagov/govuk-rfcs/blob/95b4f967a43b24141c4cd0c7feb37f3c309e21c8/rfc-091-sharing-assets.md
[slimmer]: https://github.com/alphagov/slimmer/
[slimmer_moves_scripts]: https://github.com/alphagov/slimmer/blob/0968d5b715f949cc3ef5ac3fa1dcbababd7c2fd7/lib/slimmer/processors/tag_mover.rb#L8
[specific_assets]: https://github.com/alphagov/govuk-rfcs/blob/95b4f967a43b24141c4cd0c7feb37f3c309e21c8/rfc-108-including-gem-component-assets.md
[what_slimmer_does]: https://github.com/alphagov/slimmer/blob/0968d5b715f949cc3ef5ac3fa1dcbababd7c2fd7/docs/what-slimmer-does.md#tagmover

> [^1]: Method - look at the markup present on the page. Search for all instances of `.gem-c-*` to get a list of all components being used and dedupe. Compile the Sass from the GOV.UK Publishing Components gem into individual CSS files for each component. Get the Brotli-compressed file size for each component, and add it to the application-specifc CSS. Compare this to the current CSS filesize for the application and static.

Author: 🕵️‍♂️ 👍  - fixed in https://github.com/alphagov/govuk-rfcs/pull/152/commits/35f516e8f71c2a0dc69e67367f0ac35056a6d722

### Non-rendering applications that use the components gem

There are applications that use the components gem but don't use Static and Slimmer - for example, publishing applications and the Govspeak gem.

These changes should be optional - even if this is the default behaviour, it should be able to be turned off at a component by component level and at an application level. This could be done with a helper method that checks both if the component has already been used, and if adding the component assets are allowed:

```ruby
# component_asset_helper.rb - pseudo code example:

def component_assets_allowed (component_name)
  # check:
  #  - if the component has already had the assets added to the page
  # - if the component assets loading method is allowed at an application level
  # - if the individual component has used an option to turn individual assets on or off

>   @components_in_use.include?(component_name) || ENV.component_asset_loading !== "bulk" || use_individual_assets === true

Reviewer 2: Implementation detail, but I do wonder where this `use_individual_assets` would get set, and where this helper tool should live.
We'll want to spike how this might work - but fair enough if that's out of scope for the RFC itself.

Author: I agree that this should be a spike of work - but I think that this is outside the scope of this RFC, as it's getting closer to _how_ this should be implemented rather than _why_ this should be implemented.

The applications have bespoke CSS - this varies between half a kilobyte to 15kB. This CSS can't be shared, as it's for pages only rendered by that application.

The apps also have CSS that comes from the components that they use - this varies from half a kilobyte to 16kB. This is served in the rendering applications stylesheet - so a visitor who goes from app to app can't cache and then use these assets, despite the code being the same.

### A tiny change in one component can prevent the browser using the whole of a cached JavaScript and CSS file

GOV.UK doesn't take advantage of the fact that it serves CSS and JavaScript files with a long expiry header. Because the JavaScript and CSS is concatenated and fingerprinted, a change to one component will mean the filename will change.

For example, any change to the breadcrumb component would cause a change in the CSS file for six rendering apps. Any return visitors would need to download the entire stylesheet again - serving the files individually would mean that only the updated breadcrumb component stylesheet would need to be downloaded.

> The component gem is released regularly - over the past year it's been released on average once every four days, but most releases were between one and seven days apart. Two thirds of repeat visitors to GOV.UK return within seven days and one quarter of repeat visitors return between 8 to 30 days. This means it is highly likely that returning visitors need to download the asset files again, regardless of how large or small any changes are.

Author: ```suggestion
The component gem is released regularly - over the past year it's been released on average once every four days, with most releases being between one and seven days apart. Two thirds of repeat visitors to GOV.UK return within seven days and one quarter of repeat visitors return between 8 to 30 days. This means it is highly likely that returning visitors need to download the asset files again, regardless of how large or small any changes are.
```
